File: Device/SFBSelectorControl.m
/*
 * Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

#import "SFBSelectorControl.h"
#import "SFBAudioObject+Internal.h"

@implementation SFBSelectorControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioSelectorControlClassID));
	return [super initWithAudioObjectID:objectID];
}

- (NSArray *)currentItem
{
	return [self unsignedIntArrayForProperty:kAudioSelectorControlPropertyCurrentItem inScope:SFBAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setCurrentItem:(NSArray <NSNumber *> *)values error:(NSError **)error
{
	return [self setUnsignedIntArray:values forProperty:kAudioSelectorControlPropertyCurrentItem inScope:SFBAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

- (NSArray *)availableItems
{
	return [self unsignedIntArrayForProperty:kAudioSelectorControlPropertyAvailableItems inScope:SFBAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSString *)nameOfItem:(UInt32)itemID
{
	return [self stringForProperty:kAudioSelectorControlPropertyItemName inScope:SFBAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster qualifier:&itemID qualifierSize:sizeof(itemID) error:NULL];
}

- (NSString *)kindOfItem:(UInt32)itemID
{
	return [self stringForProperty:kAudioSelectorControlPropertyItemKind inScope:SFBAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster qualifier:&itemID qualifierSize:sizeof(itemID) error:NULL];
}

@end

@implementation SFBDataSourceControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioDataSourceControlClassID));
	return [super initWithAudioObjectID:objectID];
}

@end

@implementation SFBDataDestinationControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioDataDestinationControlClassID));
	return [super initWithAudioObjectID:objectID];
}

@end

@implementation SFBClockSourceControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioClockSourceControlClassID));
	return [super initWithAudioObjectID:objectID];
}

@end

@implementation SFBLineLevelControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioLineLevelControlClassID));
	return [super initWithAudioObjectID:objectID];
}

@end

@implementation SFBHighPassFilterControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioHighPassFilterControlClassID));
	return [super initWithAudioObjectID:objectID];
}

@end

File: Device/SFBSelectorControl.swift
/*
* Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
* See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
*/

import Foundation

extension SelectorControl {
	/// Returns the selected items
	/// - note: This corresponds to `kAudioSelectorControlPropertyCurrentItem`
	/// - throws: An error if the property could not be retrieved
	public func currentItem() throws -> [UInt32] {
		return try getProperty(.selectorControlCurrentItem)
	}

	/// Returns the available items
	/// - note: This corresponds to `kAudioSelectorControlPropertyAvailableItems`
	/// - throws: An error if the property could not be retrieved
	public func availableItems() throws -> [UInt32] {
		return try getProperty(.selectorControlAvailableItems)
	}

	/// Returns the name of `itemID`
	/// - note: This corresponds to `kAudioSelectorControlPropertyItemName`
	/// - parameter itemID: The item's ID
	/// - throws: An error if the property could not be retrieved
	public func nameOfItem(_ itemID: UInt32) throws -> String {
		var qualifier = itemID
		return try __string(forProperty: .selectorControlItemName, in: .global, onElement: .master, qualifier: &qualifier, qualifierSize: UInt32(MemoryLayout<UInt32>.size))
	}

	/// Returns the kind of `itemID`
	/// - note: This corresponds to `kAudioSelectorControlPropertyItemKind`
	/// - parameter itemID: The item's ID
	/// - throws: An error if the property could not be retrieved
	public func kindOfItem(_ itemID: UInt32) throws -> String {
		var qualifier = itemID
		return try __string(forProperty: .selectorControlItemKind, in: .global, onElement: .master, qualifier: &qualifier, qualifierSize: UInt32(MemoryLayout<UInt32>.size))
	}
}
